README.md
# python-signify - OpenBSD Signify for Python

0.1.0-RC1

[![Build Status](https://travis-ci.org/bjornedstrom/python-signify.png?branch=master)](https://travis-ci.org/bjornedstrom/python-signify)

[Signify](http://www.tedunangst.com/flak/post/signify) was originally written for OpenBSD to sign files and packages, as a light-weight replacement to using PGP. **python-signify** is a module for working with Signify keys/signatures from Python. This module allow you to sign/verify messages and work with Signify keypairs.

Specifically this project contains two modules that you can use depending on requirements: the first one re-implements Signify functionality directly, and is the recommended use of python-signify. The second one uses the `subprocess` module and use the `signify` binary.

There is also a driver program using the library: `signipie`, which offer some additional convenience over the reference `signify` implementation.

#### Security Warning

Obligatory security warning for Python projects using cryptographic keys: python-signify makes **no** attempts to guard your secret keys against memory based attacks. If you use python-signify you realize that your secret keys may end up in swap space, or in computer RAM longer than necessary. If this is a problem for you, do not use python-signify.

## Installation

python-signify is tested on a few versions of Python 2 and 3.

The `signify.pure` module has a Python implementation of some parts of Signify, without requiring the `signify` binary. This module has two dependencies:

- [python-ed25519](https://github.com/warner/python-ed25519]) (`pip install ed25519`)
- [py-bcrypt](py-bcrypt) (`pip install py-bcrypt`)

If you use the `subprocess` based module, then there are no dependencies other than that `signify` is installed  on the system and reachable on PATH.

## Usage

Signify keys and signatures are `b'python bytestrings'` that start with the string `b'untrusted comment:'`. From this representation, you can create a `PublicKey`, `SecretKey` and `Signature` object using the `.from_bytes()` method, as shown in the example below (the opposite direction is called `.to_bytes()`):

### API Example (tl;dr)

```python
import signify.pure as signify

pubkey = signify.PublicKey.from_bytes(b"""untrusted comment: bjorntest public key
RWQ100QRGZoxU+Oy1g7Ko+8LjK1AQLIEavp/NuL54An1DC0U2cfCLKEl
""")

signature = signify.Signature.from_bytes(b"""untrusted comment: signature from bjorntest secret key
RWQ100QRGZoxU/gjzE8m6GYtfICqE0Ap8SdXRSHrpjnSBKMc2RMalgi5RKrEHmKfTmcsuB9ZzDCo6K6sYEqaEcEnnAFa0zCewAg=
""")

message = b"""my message
"""

print(signify.verify(pubkey, signature, message))

new_pub, new_sec = signify.generate('my new key', 'password')
new_sig = signify.sign(new_sec.unprotect('password'), message)
print(new_sig.to_bytes())
print(signify.verify(new_pub, new_sig, message))
```

### Secret Keys and Signing

Before you can use a `SecretKey` for the signing operation, you have to decrypt it using the `SecretKey.unprotect()` method. A normal pattern is as follows:

```python
from signify.pure import SecretKey, sign
sk = SecretKey.from_bytes(...)
sku = sk.unprotect('password')
sig = sign(sku, b'my message')
print(sig.to_bytes())
```

### Public Keys and Verifying

The `verify` function takes a public key, a signature and the bytestring that was signed. `InvalidSignature` will be raised on invalid signatures.

```python
from signify.pure import PublicKey, Signature, verify
pk = PublicKey.from_bytes(...)
sig = Signature.from_bytes(...)
print(verify(pk, sig, b'my message'))
```

### Generating a New Keypair

If you do not already have a Signify keypair (`signify -G`) you can generate one as follows:

```python
from signify.pure import generate
pk, sk = generate('alice aliceson', 'password')
```

The first parameter is a comment describing the key.

### Embedded Signatures

An "embedded signature" is the concatenation of a signature and the message signed. An embedded signature looks like this:

    untrusted comment: signature from signify secret key
    RWQwAARFerRo1COfT3i7SkSrTjDImrhchgmiX2Vbmy9LZdRM6j...
    Here is my signed message!

Signing and verifying embedded signatures works as follows:

```python
from signify.pure import PublicKey, SecretKey, sign, verify_embedded

# Sign
sk = SecretKey.from_bytes(...)
sku = sk.unprotect('password')
sig = sign(sku, b'my message', embed=True)
print(sig.to_bytes())

# Verify
pk = PublicKey.from_bytes(...)
embedded_signature = Signature.from_bytes(<embedded signature>)
print(verify_embedded(pk, embedded_signature))
```

### Signing and Verifying Multiple Files

A common Signify use case is to sign one or more files, such as source code distributions. The Signify convention as used in OpenBSD is to make an embedded signature (see above) of the output of OpenBSD `sha256(1)` or Linux `sha256sum --tag` (SHA512 is also common). These looks like this:

    untrusted comment: signature from signify secret key
    RWQwAARFerRo1COfT3i7SkSrTjDImrhchgmiX2Vbmy9LZdRM6jJhzMQZFLlZKKEOiEcbLAtzpvJ0TT4dqYYfClpoUfoTnnF4sgM=
    SHA256 (bjorn.pub) = a829d2df993afa575607d77ca4f1a813d9486f07f45b826386c337a2d712d721
    SHA256 (bjorn.sec) = b4b8b2d99549fa2c67848e1f6e091a3ea0696c6106755299a7424940524552c0

You sign and verify these as follows:

```python
import os
from signify.pure import PublicKey, SecretKey, Signature, sign_files, verify_files

# Sign
sk = SecretKey.from_bytes(...)
sku = sk.unprotect('password')
paths = ['bjorn.pub', 'bjorn.sec']
sig = sign_files(sku, 'SHA256', paths, root=os.getcwd())

# Verify
pk = PublicKey.from_bytes(...)
sig = Signature.from_bytes(...)
print(verify_files(pk, sig, root=os.getcwd()))
```

## Signipie

`signipie` is a command line program that is similar to `signify` but is written to be slightly more user friendly. It has the same basic functionality as `signify` but with some additional convenience helpers around key management and command line parsing.

### Signipie Key Management

Signipie will look for your key-pair(s) in `~/.signify`, where the default key-pair is `~/.signify/id_$USER` and `~/.signify/id_$USER.pub` for the secret and public key, respectively. In addition, it will look for your trusted public keys in `~/.signify/trusted/`. If you have copies of your keys in these directories, then `signipie` can be invoked without specifying key search directories.

If you have multiple Signify key-pairs, then they can be given an id instead of you having to type out full path names to the keys each time. An id is simply a name or label given to your keys. For example, the two keys `~/.signify/id_alice` and `~/.signify/id_alice.pub` collectively have the id `id_alice`.

Of course, `signipie` can be invoked with explicit keys, similar to `signify`, if that is desired.

### Signing and Verifying with Signipie

Once your file system is set up according to the layout above, you can sign and verify messages simply as follows:

    $ signipie sign my-file
    $ cat my-file.sig
	$ signipie verify -x my-file.sig my-file
	Signature Verified (key: id_bjorn.pub)

To make an embedded signature, simply add an `-e`:

    $ signipie sign -e my-file
	$ cat my-file.sig
	$ signipie verify -e -x my-file.sig

To sign multiple files:

    $ signipie sign -c bjorn.pub bjorn.sec > bjorn-files
	$ cat bjorn-files
	$ signipie verify -c -x bjorn-files

## Extras

### Subprocess based wrapper around signify(1)

`signify.wrapper` is a fairly simple (subprocess based) wrapper around the OpenBSD signify(1) command. It basically does two things for you to make your life a little bit easier:

- It handles thread safety and the actual calling of subprocess, which is easy to screw up.
- For convenience it will aid you in juggling temporary files in case you want to work with strings/buffers instead of paths, as signify(1) work on paths.

Please make sure you read and understand the library docstrings before you use the code. There are some security considerations.

The wrapper API is a little bit different from the pure API. Please consult the docstrings for more information.

### `signi.py` - Reference Driver Program

`signi.py` has similar behavior as the normal `signify` program and is included for convenience.

    $ signi.py -h
    usage:
        signi.py -C [-q] -p pubkey -x sigfile [file ...]
        signi.py -F -s seckey -x sigfile [file ...]
        signi.py -G [-n] [-c comment] -p pubkey -s seckey
        signi.py -S [-e] [-x sigfile] -s seckey -m message
        signi.py -V [-eq] [-x sigfile] -p pubkey -m message

- `-G` generates a new keypair.
- `-S` and `-V` signs and generates an arbitrary message, respectively.
- `-F` and `-C` signs and checks files, respectively (see the "Sign Files" section above for more information).

`signi.py -F` option is not included in the normal `signify` binary but is included for convenience.

## About

Copyright Björn Edström 2015. See LICENSE for details.


debian/changelog
python-signify (0.1.0-RC1) unstable; urgency=low

  * Initial Release.

 -- Björn Edström <be@bjrn.se>  Fri, 04 Sep 2015 16:55:00 +0100


test/pure_test.py
# -*- coding: utf-8 -*-
# Copyright (C) 2015 Björn Edström <be@bjrn.se>

import signify.pure as signify
import unittest


class SignifyTest(unittest.TestCase):

    KAT = [
        {
            'pub': b"""untrusted comment: bjorntest public key
RWQ100QRGZoxU+Oy1g7Ko+8LjK1AQLIEavp/NuL54An1DC0U2cfCLKEl
""",
            'priv': b"""untrusted comment: bjorntest secret key
RWRCSwAAACqHVbmAUokJcTpgKhRbw+/W+Q7nrVPi3eU100QRGZoxU86ZWb3NjEp9ScrFddFy0o2D1KtZ0440imfaWmUebGfs0Hm+Fm9SCtaJgtjFtrUlPlmnjksY8zdcXr2NvjLsr0A=
""",
            'message': b"""my message
""",
            'sig': b"""untrusted comment: signature from bjorntest secret key
RWQ100QRGZoxU/gjzE8m6GYtfICqE0Ap8SdXRSHrpjnSBKMc2RMalgi5RKrEHmKfTmcsuB9ZzDCo6K6sYEqaEcEnnAFa0zCewAg=
""",
            'brokensig': b"""untrusted comment: signature from bjorntest secret key
RWQ100QRGZoxU/gjzE8m6GYtfICqE0Ap8SdXRSHrpjnSBKMc2RMXlgi5RKrEHmKfTmcsuB9ZzDCo6K6sYEqaEcEnnAFa0zCewAg=
""",
            'embedded': b"""untrusted comment: signature from bjorntest secret key
RWQ100QRGZoxU/gjzE8m6GYtfICqE0Ap8SdXRSHrpjnSBKMc2RMalgi5RKrEHmKfTmcsuB9ZzDCo6K6sYEqaEcEnnAFa0zCewAg=
my message
"""
            }
        ]

    def test_extraction(self):
        self.assertEquals(
            b'\xe3\xb2\xd6\x0e\xca\xa3\xef\x0b\x8c\xad@@\xb2\x04j\xfa\x7f6\xe2\xf9\xe0\t\xf5\x0c-\x14\xd9\xc7\xc2,\xa1%',
            signify.PublicKey.from_bytes(self.KAT[0]['pub']).raw())

        sk = signify.SecretKey.from_bytes(self.KAT[0]['priv'])
        sku = sk.unprotect('test')
        self.assertEquals(
            b'D@\xd9\xca\xb2\x96;\xa0^\xbb\x16\xc8\x0f\xf7Y=(hu\x85\xbd\xe4i\xf6\xcf\x0f\xfb#\xc1\xfa\xe0\xa1\xe3\xb2\xd6\x0e\xca\xa3\xef\x0b\x8c\xad@@\xb2\x04j\xfa\x7f6\xe2\xf9\xe0\t\xf5\x0c-\x14\xd9\xc7\xc2,\xa1%',
            sku.raw_secret_key())

    def test_verify_success(self):
        self.assertTrue(
            signify.verify(signify.PublicKey.from_bytes(self.KAT[0]['pub']),
                           signify.Signature.from_bytes(self.KAT[0]['sig']),
                           self.KAT[0]['message']))

    def test_sign(self):
        sk = signify.SecretKey.from_bytes(self.KAT[0]['priv'])
        sku = sk.unprotect('test')
        sig = signify.sign(sku,
                           self.KAT[0]['message'])

        self.assertEquals(self.KAT[0]['sig'], sig.to_bytes())

    def test_sign_embedded(self):
        sk = signify.SecretKey.from_bytes(self.KAT[0]['priv'])
        sku = sk.unprotect('test')
        sig = signify.sign(sku,
                           self.KAT[0]['message'],
                           True)

        self.assertEquals(self.KAT[0]['embedded'], sig.to_bytes())

    def test_verify_embedded(self):
        self.assertTrue(
            signify.verify_embedded(signify.PublicKey.from_bytes(self.KAT[0]['pub']),
                                    self.KAT[0]['embedded']))

        self.assertEquals(b'my message\n',
            signify.verify_embedded(signify.PublicKey.from_bytes(self.KAT[0]['pub']),
                                    signify.Signature.from_bytes(self.KAT[0]['embedded'])))

    def test_decrypt_secret_wrong_password(self):
        self.assertRaises(KeyError,
                          signify.SecretKey.from_bytes(self.KAT[0]['priv']).unprotect,
                          'wrongpassword')

    def test_verify_failure(self):
        self.assertRaises(
            signify.InvalidSignature,
            signify.verify, signify.PublicKey.from_bytes(self.KAT[0]['pub']),
                            signify.Signature.from_bytes(self.KAT[0]['brokensig']),
                            self.KAT[0]['message'])

    def test_generate_sign_no_password(self):
        pub, priv = signify.generate('test', None)

        self.assertTrue(pub.to_bytes().startswith(b'untrusted comment: test public key'))
        self.assertTrue(priv.to_bytes().startswith(b'untrusted comment: test secret key'))

        sku = priv.unprotect(None)
        sig = signify.sign(sku,
                           b'My Message')

        self.assertTrue(
            signify.verify(pub,
                           sig,
                           b'My Message'))

    def test_generate_no_comment(self):
        pub, priv = signify.generate(None, None)

        self.assertTrue(pub.to_bytes().startswith(b'untrusted comment: signify public key'))
        self.assertTrue(priv.to_bytes().startswith(b'untrusted comment: signify secret key'))

    def test_generate_sign_with_password(self):
        pub, priv = signify.generate(None, 'testpassword')

        self.assertTrue(pub.to_bytes().startswith(b'untrusted comment: signify public key'))
        self.assertTrue(priv.to_bytes().startswith(b'untrusted comment: signify secret key'))

        sku = priv.unprotect('testpassword')
        sig = signify.sign(sku,
                           b'My Message')

        self.assertTrue(
            signify.verify(pub,
                           sig,
                           b'My Message'))


if __name__ == '__main__':
    unittest.main()


test/system_test.py
# -*- coding: utf-8 -*-
# Copyright (C) 2015 Björn Edström <be@bjrn.se>

import signify.wrapper as signify
import unittest


class SignifyTest(unittest.TestCase):

    KAT = [
        {
            'pub': b"""untrusted comment: bjorntest public key
RWQ100QRGZoxU+Oy1g7Ko+8LjK1AQLIEavp/NuL54An1DC0U2cfCLKEl
""",
            'priv': b"""untrusted comment: bjorntest secret key
RWRCSwAAACqHVbmAUokJcTpgKhRbw+/W+Q7nrVPi3eU100QRGZoxU86ZWb3NjEp9ScrFddFy0o2D1KtZ0440imfaWmUebGfs0Hm+Fm9SCtaJgtjFtrUlPlmnjksY8zdcXr2NvjLsr0A=
""",
            'message': b"""my message
""",
            'sig': b"""untrusted comment: signature from bjorntest secret key
RWQ100QRGZoxU/gjzE8m6GYtfICqE0Ap8SdXRSHrpjnSBKMc2RMalgi5RKrEHmKfTmcsuB9ZzDCo6K6sYEqaEcEnnAFa0zCewAg=
"""
            }
        ]

    def setUp(self):
        self.obj = signify.Signify()

    def test_verify_success(self):
        self.assertTrue(
            self.obj.verify_simple(self.KAT[0]['pub'],
                                   self.KAT[0]['sig'],
                                   self.KAT[0]['message']))

    def test_verify_failure(self):
        broken_sig = b"""untrusted comment: signature from bjorntest secret key
RWQ100QRGZoxU/gjzE8m6GYtfICqE0Ap8SdXRSHrpjnSBKMc2RMagic5RKrEHmKfTmcsuB9ZzDCo6K6sYEqaEcEnnAFa0zCewAg=
"""

        self.assertRaises(
            signify.InvalidSignature,
            self.obj.verify_simple, self.KAT[0]['pub'],
                                    broken_sig,
                                    self.KAT[0]['message'])

    def test_generate(self):
        pub, priv = self.obj.generate_unsafe('test', None)

    def test_sign(self):
        pub, priv = self.obj.generate_unsafe('test', None)

        sig = self.obj.sign(privkey_buf=priv,
                            message_buf=b'My Message')

        self.assertTrue(
            self.obj.verify_simple(pub,
                                   sig,
                                   b'My Message'))

if __name__ == '__main__':
    unittest.main()


test/signipie-integration-test.sh
#!/bin/bash

export PYTHONPATH=.
export PATH=bin:$PATH
HOMEDIR=$(mktemp -d)
PLAYDIR=$(mktemp -d)

echo $HOMEDIR
echo $PLAYDIR

# generate
signipie --home "$HOMEDIR" -i id_alice generate -n
echo $?
signipie --home "$HOMEDIR" -i id_bob generate -n
echo $?
signipie --home "$HOMEDIR" -i id_mallory generate -n
echo $?
mkdir "$HOMEDIR/trusted"
cp "$HOMEDIR/id_alice.pub" "$HOMEDIR/trusted"
cp "$HOMEDIR/id_bob.pub" "$HOMEDIR/trusted"

# files for signing and stuff
echo 'My simple message to sign' > "$PLAYDIR/simple"
echo 'A little file 1' > "$PLAYDIR/file1"
echo 'A little file 2' > "$PLAYDIR/file2"
echo 'A little file 3' > "$PLAYDIR/file3"

# sign and verify
echo "SIMPLE SIGNATURE"
echo "----------------"
signipie --home "$HOMEDIR" -i id_alice sign "$PLAYDIR/simple"
echo $?
cat "$PLAYDIR/simple.sig"
signify -S -x "$PLAYDIR/simple.sig-ref" -s "$HOMEDIR/id_alice" -m "$PLAYDIR/simple"
cat "$PLAYDIR/simple.sig-ref"

signipie --home "$HOMEDIR" verify -x "$PLAYDIR/simple.sig" "$PLAYDIR/simple"
echo $?
echo
echo "EMBEDDED SIGNATURE"
echo "------------------"
signipie --home "$HOMEDIR" -i id_alice sign -e "$PLAYDIR/simple" -o "$PLAYDIR/simple.embedded"
echo $?
cat "$PLAYDIR/simple.embedded"
signify -S -e -x "$PLAYDIR/simple.embedded-ref" -s "$HOMEDIR/id_alice" -m "$PLAYDIR/simple"
cat "$PLAYDIR/simple.embedded-ref"

signipie --home "$HOMEDIR" verify -e "$PLAYDIR/simple.embedded"
echo $?
echo
echo "CHECKSUM SIGNATURES"
echo "-------------------"
signipie --home "$HOMEDIR" -i id_alice sign -c -o "$PLAYDIR/simple.checksum" "$PLAYDIR/file1" "$PLAYDIR/file2" "$PLAYDIR/file3"
echo $?
cat "$PLAYDIR/simple.checksum"
sha256sum --tag "$PLAYDIR/file1" "$PLAYDIR/file2" "$PLAYDIR/file3" > "$PLAYDIR/fileN.hashes"
signify -S -e -x "$PLAYDIR/simple.checksum-ref" -s "$HOMEDIR/id_alice" -m "$PLAYDIR/fileN.hashes"
cat "$PLAYDIR/simple.checksum-ref"

signipie --home "$HOMEDIR" verify -c -x "$PLAYDIR/simple.checksum"
echo $?
echo
echo "TEST FAILURES"
echo "-------------"

signipie --home "$HOMEDIR" -p "$HOMEDIR/id_bob.pub" verify -x "$PLAYDIR/simple.sig" "$PLAYDIR/simple"
echo $?
signipie --home "$HOMEDIR" -p "$HOMEDIR/id_bob.pub" verify -e "$PLAYDIR/simple.embedded"
echo $?
signipie --home "$HOMEDIR" -p "$HOMEDIR/id_bob.pub" verify -c -x "$PLAYDIR/simple.checksum"
echo $?

echo "broken" >> "$PLAYDIR/file2"
rm "$PLAYDIR/file3"
signipie --home "$HOMEDIR" verify -c -x "$PLAYDIR/simple.checksum"
echo $?
signify -C -p "$HOMEDIR/id_alice.pub" -x "$PLAYDIR/simple.checksum"


# Cleanup
rm -r "$HOMEDIR"
rm -r "$PLAYDIR"


signify/check.py
# -*- coding: utf-8 -*-
# Copyright (c) Björn Edström <be@bjrn.se> 2015. See LICENSE for details.

import hashlib
import re
import os

BUFSIZE = 64*1024
ALGO_TO_CLS = {
    'SHA256': hashlib.sha256,
    'SHA512': hashlib.sha512,
}


def hash_file(hashobj, path):
    with open(path, 'rb') as fobj:
        while True:
            buf = fobj.read(BUFSIZE)
            hashobj.update(buf)
            if buf != BUFSIZE:
                break
    return True


def openbsd_sha_files(algo_str, root, files):
    hash_cls = ALGO_TO_CLS[algo_str]
    res = []
    for path in files:
        hashobj = hash_cls()
        hash_file(hashobj, os.path.join(root, path))
        res.append('%s (%s) = %s' % (algo_str, path, hashobj.hexdigest()))
    return '\n'.join(res) + '\n'


def checkfiles(root, checkfile):
    for algo, path, ref_digest in re.findall(
            r'^(\S+) [(]([^)]+)[)] = ([0-9a-fA-F]+)$', checkfile, re.M):
        hash_cls = ALGO_TO_CLS[algo]
        hashobj = hash_cls()
        try:
            hash_file(hashobj, os.path.join(root, path))
            digest = hashobj.hexdigest()
            status = digest.upper() == ref_digest.upper()
        except Exception as e:
            status = e
        yield (path, status)


signify/pure.py
# -*- coding: utf-8 -*-
# Copyright (c) Björn Edström <be@bjrn.se> 2015. See LICENSE for details.

"""Signify library

This library provide a few main methods and some minor helpers:

    - sign(...)
    - verify(...)
    - generate(...)

In addition it provide 3 helper classes:

    - PublicKey
    - SecretKey
    - Signature

A signify key/signature is a byte string on the form:
"untrusted comment: ...\\n<base64 blob>\\n"

Use the {PublicKey, SecretKey, Signature}.from_bytes(...) method to
create an object instance from a byte string of the above form. Use
the .to_bytes() method for the opposite direction:

    >>> pubkey = PublicKey.from_bytes(b'''untrusted comment: bjorntest public key
RWQ100QRGZoxU+Oy1g7Ko+8LjK1AQLIEavp/NuL54An1DC0U2cfCLKEl
''')
    >>> print(pubkey.to_bytes())
    b'untrusted...'
    >>> # same goes for Signature and SecretKey

SecretKey:s may be password protected. Use the .unprotect(...) method
to get an UnprotectedSecretKey, which can be used by sign(...).

Example
-------

    pubkey = signify.PublicKey.from_bytes(b'''untrusted comment: bjorntest public key
    RWQ100QRGZoxU+Oy1g7Ko+8LjK1AQLIEavp/NuL54An1DC0U2cfCLKEl
    ''')

    signature = signify.Signature.from_bytes(b'''untrusted comment: signature from bjorntest secret key
    RWQ100QRGZoxU/gjzE8m6GYtfICqE0Ap8SdXRSHrpjnSBKMc2RMalgi5RKrEHmKfTmcsuB9ZzDCo6K6sYEqaEcEnnAFa0zCewAg=
    ''')

    message = b'''my message
    '''

    print(signify.verify(pubkey, signature, message))

    new_pub, new_sec = signify.generate('my new key', 'password')
    new_sig = signify.sign(new_sec.unprotect('password'), message)
    print(new_sig.to_bytes())
    print(signify.verify(new_pub, new_sig, message))

Security Warning
----------------

Python is an interpreted language and there are no guarantees that
your signify secret key, ones unprotected, may end up in swap space or
be recoverable from reading your computers memory.

"""

import base64
import bcrypt
import ed25519
import hashlib
import os
import re
import struct

import signify.check as check
from signify.util import *


class SignifyError(Exception):
    pass


class InvalidSignature(SignifyError):
    pass


class _Materialized(object):
    def __init__(self):
        self._blob = None
        self._comment = None
        self._raw = None
        self._keynum = None

    def comment(self):
        """Returns the plain text comment associated with the object."""
        return self._comment

    def keynum(self):
        """Returns the 8 byte identifier for the object."""
        return self._keynum

    def raw(self):
        return self._raw

    def to_bytes(self):
        """Returns the materialized byte blob for the object."""
        return self._blob

    @staticmethod
    def write_message(comment, blob):
        assert isinstance(comment, (str, unicode))
        assert isinstance(blob, bytes)
        return b'untrusted comment: ' + comment.encode('utf-8') + b'\n' + base64.b64encode(blob) + b'\n'

    @staticmethod
    def read_message(msg):
        assert isinstance(msg, bytes)

        try:
            comment_line, b64 = msg.split(b'\n', 1)
        except:
            raise SignifyError('malformed message')

        try:
            blob = base64.b64decode(b64)
        except:
            raise SignifyError('malformed message: base64 error')

        try:
            comment = re.findall(r'^untrusted comment: (.*?)$', comment_line.decode('utf-8'))[0]
        except:
            raise SignifyError('malformed message: expected a comment line')

        return (comment, bytes(blob))


class Signature(_Materialized):
    """A Signature object."""

    def __init__(self):
        _Materialized.__init__(self)

    def _parse_sigfile(self, msg):
        self._comment, blob = _Materialized.read_message(msg)

        try:
            pkalg, keynum, sig = \
                struct.unpack('!2s8s64s', blob)

            assert pkalg == b'Ed'
        except Exception as e:
            raise SignifyError('malformed signature blob')

        return sig, keynum

    @staticmethod
    def from_bytes(blob):
        assert isinstance(blob, bytes)
        obj = Signature()
        obj._raw, obj._keynum = obj._parse_sigfile(blob)
        obj._blob = blob
        return obj

    def __repr__(self):
        return '<Signature by %s>' % (bytes2hex(self._keynum))


class PublicKey(_Materialized):
    """A Signify public key."""

    def __init__(self):
        _Materialized.__init__(self)

    def _parse_public_key(self, msg):
        self._comment, blob = _Materialized.read_message(msg)

        try:
            pkalg, keynum, pubkey = \
                struct.unpack('!2s8s32s', blob)

            assert pkalg == b'Ed'
        except Exception as e:
            raise SignifyError('malformed public key')

        return pubkey, keynum

    @staticmethod
    def from_bytes(blob):
        assert isinstance(blob, bytes)
        obj = PublicKey()
        obj._blob = blob
        obj._raw, obj._keynum = obj._parse_public_key(blob)
        return obj

    def __repr__(self):
        return '<PublicKey %s>' % (bytes2hex(self._keynum))


class UnprotectedSecretKey(object):
    """This class represents a decrypted secret key that can be used for
    signing operations."""

    def __init__(self, sk):
        assert isinstance(sk, SecretKey)
        self._sk = sk
        self._key = None

    def comment(self):
        return self._sk.comment()

    def keynum(self):
        return self._sk.keynum()

    def raw_secret_key(self):
        return self._key


class SecretKey(_Materialized):
    """A Signify secret key object."""

    def __init__(self):
        _Materialized.__init__(self)

    @staticmethod
    def from_bytes(blob):
        assert isinstance(blob, bytes)
        obj = SecretKey()
        obj._blob = blob
        obj._parse_secret_key(blob)
        return obj

    def _parse_secret_key(self, msg):
        comment, blob = _Materialized.read_message(msg)

        try:
            pkalg, kdfalg, kdfrounds, salt, checksum, keynum, seckey = \
                struct.unpack(b'!2s2sL16s8s8s64s', blob)

            assert pkalg == b'Ed'
            assert kdfalg == b'BK'
            assert kdfrounds in [0, 42]
        except Exception as e:
            raise SignifyError(e)

        self._kdfrounds = kdfrounds
        self._salt = salt
        self._checksum = checksum
        self._keynum = keynum
        self._seckey = seckey
        self._comment = comment

    def is_password_protected(self):
        """Returns True if this secret key is protected with a password."""

        return self._kdfrounds != 0

    def raw(self):
        raise NotImplementedError('call unprotect()')

    def unprotect(self, password):
        """Decrypt the SecretKey object and return an UnprotectedSecretKey
        that can be used for signing.

        Can throw KeyError if the password is incorrect.
        """

        if self._kdfrounds == 0:
            xorkey = b'\x00' * 64
        else:
            xorkey = bcrypt.kdf(password, self._salt, 64, self._kdfrounds)
        priv = xorbuf(self._seckey, xorkey)

        checksum_ref = hashlib.sha512(priv).digest()[0:8]
        if self._checksum != checksum_ref:
            raise KeyError('incorrect password')

        usk = UnprotectedSecretKey(self)
        usk._key = priv
        return usk

    def __repr__(self):
        return '<SecretKey %s>' % (bytes2hex(self._keynum))


def sign(secret_key, message, embed=False):
    """Sign a message with the secret key.

    @param secret_key: The secret key to sign the message with.
    @param message: The message to sign.
    @param embed: Whether to create an embedded signature or not.
    """

    assert isinstance(secret_key, UnprotectedSecretKey)
    assert isinstance(message, bytes)

    key_obj = ed25519.keys.SigningKey(secret_key.raw_secret_key())

    sig_buf = key_obj.sign(message)

    sig_blob = b'Ed' + secret_key.keynum() + sig_buf

    sig = _Materialized.write_message('signature from %s' % (secret_key.comment(),), sig_blob)
    if embed:
        sig += message
    return Signature.from_bytes(sig)


def verify(public_key, signature, message):
    """Verify a signature.

    Will return True on success or raise InvalidSignature.

    @param public_key: The public key corresponding to the secret key
    that signed the message.
    @param signature: The signature.
    @param message: The message signed.
    """

    assert isinstance(public_key, PublicKey)
    assert isinstance(signature, Signature)
    assert isinstance(message, bytes)

    key_obj = ed25519.keys.VerifyingKey(public_key.raw())
    try:
        key_obj.verify(signature.raw(), message)
        return True
    except ed25519.BadSignatureError as e:
        raise InvalidSignature('signify: signature verification failed')


def verify_embedded(public_key, embedded_message):
    """Verify an embedded signature.

    If the signature verification fails, raise InvalidSignature.

    Otherwise, it will return the message embedded in the signature.

    @param public_key: The public key corresponding to the secret key
    that signed the message.
    @param embedded_message: The message(bytes) or Signature.
    """

    assert isinstance(public_key, PublicKey)
    assert isinstance(embedded_message, (bytes, Signature))

    if isinstance(embedded_message, Signature):
        embedded_message = embedded_message.to_bytes()

    sig1, sig2, message = embedded_message.split(b'\n', 2)
    sig = sig1 + b'\n' + sig2 + b'\n'

    verify(public_key, Signature.from_bytes(sig), message)

    return message


def generate_from_raw(comment, password, raw_pub, raw_priv):
    """ADVANCED: Given a raw Ed25519 key pair raw_pub and raw_priv,
    create a Signify keypair.

    See generate() for documentation.
    """

    assert isinstance(raw_pub, bytes)
    assert isinstance(raw_priv, bytes)

    if comment is None:
        comment = 'signify'

    keynum = os.urandom(8)

    # private key
    kdfrounds = 42
    salt = os.urandom(16)

    if password is None:
        kdfrounds = 0
        xorkey = b'\x00' * 64
    else:
        xorkey = bcrypt.kdf(password, salt, 64, kdfrounds)
    protected_key = xorbuf(xorkey, raw_priv)
    checksum = hashlib.sha512(raw_priv).digest()[0:8]

    priv_blob = b'Ed' + b'BK' + struct.pack('!L', kdfrounds) + \
                salt + checksum + keynum + protected_key
    priv = _Materialized.write_message('%s secret key' % (comment,), priv_blob)

    # public key
    pub_blob = b'Ed' + keynum + raw_pub
    pub = _Materialized.write_message('%s public key' % (comment,), pub_blob)

    return PublicKey.from_bytes(pub), SecretKey.from_bytes(priv)


def generate(comment, password):
    """Generate a signify keypair.

    @param comment: A comment to name the keypair, or None.
    @param password: A password to protect the private key, or None.
    """

    assert isinstance(comment, (type(None), str, unicode))
    #assert isinstance(password, bytes)

    sk, vk = ed25519.keys.create_keypair()
    return generate_from_raw(comment, password, vk.to_bytes(), sk.to_bytes())


def sign_files(secret_key, algo, paths, root='.'):
    """Make an embedded signature containing files and file hashes.

    This is equivlant to signing the output of `sha256sum --tag` (or
    `sha512sum`) and is provided here for portability.

    @param secret_key: The secret key to sign with.
    @param algo: Either the string 'SHA256' or 'SHA512'.
    @param paths: A list of paths to hash and sign.
    @param root: The root directory `paths` are relative from.
    """

    msg = check.openbsd_sha_files(algo, root, paths).encode('utf-8')
    return sign(secret_key, msg, embed=True)


def verify_files(public_key, signature, root='.'):
    """Verify the output of sign_files().

    Raise InvalidSignature on error.

    Returns a list of (path, status) tuples where status is True if
    checksum matches, otherwise False if verification fails, or an
    Exception instance (if for example the file can't be opened).

    @param public_key: The public key to check against.
    @param signature: The embedded signature containing the hashes.
    @param root: The directory of which the paths in the signature are
    relative against.
    """

    checkfile = verify_embedded(public_key, signature)

    return list(check.checkfiles(root, checkfile.decode('utf-8')))
